sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/routing/LocationCache.java

```diff
@@ -129,52 +133,95 @@ public void onDatabaseAccountRead(DatabaseAccount databaseAccount) {
 
     void onLocationPreferenceChanged(UnmodifiableList<String> preferredLocations) {
         this.updateLocationCache(
-                null, null , preferredLocations, null);
+                null, null, preferredLocations, null);
+    }
+
+    /*
+     * Determines whether a request is a meta-data request.
+     */
+    public boolean isMetaData(RxDocumentServiceRequest request) {
+        return (request.getOperationType() != com.azure.cosmos.implementation.OperationType.ExecuteJavaScript
+                && request.getResourceType() == ResourceType.StoredProcedure)
+                || request.getResourceType() != ResourceType.Document;
+    }
+
+    /*
+     * Determines whether a request is a meta-data write request on a multimaster
+     * account
+     */
+    public boolean isMultimasterMetadataWriteRequest(RxDocumentServiceRequest request) {
+        return !request.isReadOnlyRequest() && this.locationInfo.availableWriteLocations.size() > 1
+                && this.isMetaData(request)
+                && this.canUseMultipleWriteLocations();
+    }
+
+    /*
+     * Returns hub URI
+     */
+    public URI getHubUri() {
+        DatabaseAccountLocationsInfo currentLocationInfo = this.locationInfo;
+        String writeLocation = currentLocationInfo.availableWriteLocations.get(0);
+        return currentLocationInfo.availableWriteEndpointByLocation.get(writeLocation);
     }
 
     /**
      * Resolves request to service endpoint.
      * 1. If this is a write request
-     *    (a) If UseMultipleWriteLocations = true
-     *        (i) For document writes, resolve to most preferred and available write endpoint.
-     *            Once the endpoint is marked unavailable, it is moved to the end of available write endpoint. Current request will
-     *            be retried on next preferred available write endpoint.
-     *        (ii) For all other resources, always resolve to first/second (regardless of preferred locations)
-     *             write getEndpoint in {@link DatabaseAccount#getWritableLocations()}.
-     *             Endpoint of first write location in {@link DatabaseAccount#getWritableLocations()} is the only getEndpoint that supports
-     *             write operation on all resource types (except during that region's failover).
-     *             Only during manual failover, client would retry write on second write location in {@link DatabaseAccount#getWritableLocations()}.
-     *    (b) Else resolve the request to first write getEndpoint in {@link DatabaseAccount#getWritableLocations()} OR
-     *        second write getEndpoint in {@link DatabaseAccount#getWritableLocations()} in case of manual failover of that location.
-     * 2. Else resolve the request to most preferred available read getEndpoint (getAutomatic failover for read requests)
+     * (a) If UseMultipleWriteLocations = true
```

Comment:
I like the previously format better -- it is easier to read, can we revert back? 

( side notes: there are a lot style related changes in this PR which does not related to the logic change, it might be better to change in a different PR, helps reduce the scope)